codec-dns-doh/src/main/java/io/netty/handler/codec/dns/doh/DoHEncoder.java

```diff
@@ -0,0 +1,128 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns.doh;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToMessageEncoder;
+import io.netty.handler.codec.base64.Base64;
+import io.netty.handler.codec.base64.Base64Dialect;
+import io.netty.handler.codec.dns.DnsQuery;
+import io.netty.handler.codec.dns.DnsRecordEncoder;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.util.CharsetUtil;
+
+import java.net.URL;
+import java.util.List;
+
+/**
+ * Encodes {@link DnsQuery} into {@link FullHttpRequest}
+ */
+public final class DoHEncoder extends MessageToMessageEncoder<DnsQuery> {
+
+    private final DnsQueryEncoder encoder;
+    private final boolean http2;
+    private final boolean get;
+    private final URL url;
+
+    /**
+     * Creates a new encoder which uses {@linkplain DnsRecordEncoder#DEFAULT the default record encoder},
+     * HTTP/2 (h2) and HTTP GET method by default.
+     *
+     * @param url DoH Upstream Server
+     * @throws IllegalArgumentException If URL Protocol is not HTTPS
+     */
+    public DoHEncoder(URL url) throws IllegalArgumentException {
+        this(DnsRecordEncoder.DEFAULT, true, true, url);
+    }
+
+    /**
+     * Creates a new encoder which uses {@linkplain DnsRecordEncoder#DEFAULT the default record encoder},
+     * and HTTP/2 (h2) by default.
+     *
+     * @param httpGET Use HTTP GET method
+     * @param url     DoH Upstream Server
+     * @throws IllegalArgumentException If URL Protocol is not HTTPS
+     */
+    public DoHEncoder(boolean httpGET, URL url) throws IllegalArgumentException {
+        this(DnsRecordEncoder.DEFAULT, true, httpGET, url);
+    }
+
+    /**
+     * Creates a new encoder with the specified {@code recordEncoder}, {@code HTTP2},
+     * {@code httpGET} and {@code url}
+     *
+     * @param recordEncoder DNS Record Encoder
+     * @param useHTTP2      Use HTTP/2 (h2)
+     * @param httpGET       Use HTTP GET method
+     * @param url           DoH Upstream Server
+     * @throws IllegalArgumentException If URL Protocol is not HTTPS
+     */
+    public DoHEncoder(DnsRecordEncoder recordEncoder, boolean useHTTP2, boolean httpGET, URL url)
+            throws IllegalArgumentException {
+        this.encoder = new DnsQueryEncoder(recordEncoder);
+        this.http2 = useHTTP2;
+        this.get = httpGET;
+        this.url = url;
+
+        // DoH queries must be done on top of HTTPS (HTTP + TLS)
+        // Reference: https://tools.ietf.org/html/rfc8484#section-5
+        if (!url.getProtocol().equalsIgnoreCase("https")) {
+            throw new IllegalArgumentException("Protocol is not HTTPS");
```

Comment:
TIOLI: Useful to include url at this point so an unexpected value can be traced back.

Reply:
Done!